Build fails with Kotlin coroutine/suspend methods

### Describe the bug

With latest release the build fails with this error message:
```
Build step io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor#setupEndpoints threw an exception: java.lang.RuntimeException: java.lang.RuntimeException: Failed to process method 'io.outfoxx.cloud.api.admin.server.AdminAPI#createAccount'
```

and the final cause error message is:
```
Resource method java.lang.Object createAccount(java.lang.String tenantId, io.outfoxx.cloud.api.CreateAccountParams body, javax.ws.rs.core.UriInfo uriInfo, kotlin.coroutines.Continuation<? super javax.ws.rs.core.Response> $completion) can only have a single body parameter: $completion
```

Notice the complaint about the `$completion` parameter in the method.
 

The suspend method is defined in an interface with the following signature:
```kotlin
@Path(value = "/admin/api/v1.0")
@Produces(value = ["application/json","application/cbor"])
@Consumes(value = ["application/json"])
interface AdminAPI
  
  //...

  @POST
  @Path(value = "/tenants/{tenantId}/accounts")
  public suspend fun createAccount(
    @PathParam(value = "tenantId") tenantId: String,
    @Valid body: CreateAccountParams,
    @Context uriInfo: UriInfo
  ): Response

  //...

}
```

An implementation class then derives from the interface and implements the methods. It builds fine in `2.13.0.Final`

### Expected behavior

Kotlin suspend methods work as they did in previous releases.

### Actual behavior

Kotlin suspend methods in interfaces fails the 

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

macOS 12.6

### Output of `java -version`

OpenJDK 17.0.1

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.13.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5

### Additional information

_No response_

Here is the entire stack trace, it references `EndpointIndexer.createEndpoints`.

<details>

```
java.lang.RuntimeException: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor#setupEndpoints threw an exception: java.lang.RuntimeException: java.lang.RuntimeException: Failed to process method 'io.outfoxx.cloud.api.admin.server.AdminAPI#createAccount'
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:315)
        at io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor.setupEndpoints(ResteasyReactiveProcessor.java:593)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
  Caused by: java.lang.RuntimeException: Failed to process method 'io.outfoxx.cloud.api.admin.server.AdminAPI#createAccount'
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:703)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:405)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:452)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:286)
        ... 12 more
  Caused by: java.lang.RuntimeException: Resource method java.lang.Object createAccount(java.lang.String tenantId, io.outfoxx.cloud.api.CreateAccountParams body, javax.ws.rs.core.UriInfo uriInfo, kotlin.coroutines.Continuation<? super javax.ws.rs.core.Response> $completion) can only have a single body parameter: $completion
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:551)
        ... 15 more
```

</details>

--- a/devtools/gradle/gradle-application-plugin/build.gradle
+++ b/devtools/gradle/gradle-application-plugin/build.gradle
@@ -33,7 +33,3 @@ pluginBundle {
     vcsUrl = 'https://github.com/quarkusio/quarkus'
     tags = ['quarkus', 'quarkusio', 'graalvm']
 }
-
-test {
-    systemProperty 'kotlin_version', project.kotlin_version
-}

--- a/devtools/gradle/gradle-application-plugin/build.gradle
+++ b/devtools/gradle/gradle-application-plugin/build.gradle
@@ -33,3 +33,7 @@ pluginBundle {
     vcsUrl = 'https://github.com/quarkusio/quarkus'
     tags = ['quarkus', 'quarkusio', 'graalvm']
 }
+
+test {
+    systemProperty 'kotlin_version', project.kotlin_version
+}

--- a/devtools/gradle/gradle-application-plugin/pom.xml
+++ b/devtools/gradle/gradle-application-plugin/pom.xml
@@ -54,6 +54,18 @@
             <artifactId>quarkus-devmode-test-utils</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-gradle-plugin</artifactId>
+            <version>${kotlin.version}</version>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.checkerframework</groupId>
+                    <artifactId>checker-qual</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
     </dependencies>
 
     <build>

--- a/devtools/gradle/gradle-application-plugin/pom.xml
+++ b/devtools/gradle/gradle-application-plugin/pom.xml
@@ -54,18 +54,6 @@
             <artifactId>quarkus-devmode-test-utils</artifactId>
             <scope>test</scope>
         </dependency>
-        <dependency>
-            <groupId>org.jetbrains.kotlin</groupId>
-            <artifactId>kotlin-gradle-plugin</artifactId>
-            <version>${kotlin.version}</version>
-            <scope>test</scope>
-            <exclusions>
-                <exclusion>
-                    <groupId>org.checkerframework</groupId>
-                    <artifactId>checker-qual</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
     </dependencies>
 
     <build>

--- a/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/QuarkusPluginTest.java
+++ b/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/QuarkusPluginTest.java
@@ -1,15 +1,10 @@
 package io.quarkus.gradle;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.gradle.testkit.runner.TaskOutcome.SUCCESS;
-import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.io.File;
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
@@ -21,17 +16,11 @@ import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.TaskContainer;
 import org.gradle.testfixtures.ProjectBuilder;
-import org.gradle.testkit.runner.BuildResult;
-import org.gradle.testkit.runner.GradleRunner;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.io.TempDir;
 
 import io.quarkus.gradle.extension.QuarkusPluginExtension;
 
 public class QuarkusPluginTest {
-
-    @TempDir
-    Path testProjectDir;
 
     @Test
     public void shouldCreateTasks() {
@@ -78,7 +67,7 @@ public class QuarkusPluginTest {
     }
 
     @Test
-    public void shouldReturnMultipleOutputSourceDirectories() {
+    public void shouldReturnMutlipleOutputSourceDirectories() {
         Project project = ProjectBuilder.builder().build();
         project.getPluginManager().apply(QuarkusPlugin.ID);
         project.getPluginManager().apply("java");
@@ -95,68 +84,7 @@ public class QuarkusPluginTest {
 
     }
 
-    @Test
-    public void shouldNotFailOnProjectDependenciesWithoutMain() throws IOException {
-        var kotlinVersion = System.getProperty("kotlin_version", "1.7.20");
-        var settingFile = testProjectDir.resolve("settings.gradle.kts");
-        var mppProjectDir = testProjectDir.resolve("mpp");
-        var quarkusProjectDir = testProjectDir.resolve("quarkus");
-        var mppBuild = mppProjectDir.resolve("build.gradle.kts");
-        var quarkusBuild = quarkusProjectDir.resolve("build.gradle.kts");
-        Files.createDirectory(mppProjectDir);
-        Files.createDirectory(quarkusProjectDir);
-        Files.writeString(settingFile,
-                "rootProject.name = \"quarkus-mpp-sample\"\n" +
-                        "\n" +
-                        "include(\n" +
-                        "    \"mpp\",\n" +
-                        "    \"quarkus\"\n" +
-                        ")");
-
-        Files.writeString(mppBuild,
-                "buildscript {\n" +
-                        "    repositories {\n" +
-                        "        mavenLocal()\n" +
-                        "        mavenCentral()\n" +
-                        "    }\n" +
-                        "    dependencies {\n" +
-                        "        classpath(\"org.jetbrains.kotlin:kotlin-gradle-plugin:" + kotlinVersion + "\")\n" +
-                        "    }\n" +
-                        "}\n" +
-                        "\n" +
-                        "apply(plugin = \"org.jetbrains.kotlin.multiplatform\")\n" +
-                        "\n" +
-                        "repositories {\n" +
-                        "    mavenCentral()\n" +
-                        "}\n" +
-                        "\n" +
-                        "configure<org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension>{\n" +
-                        "    jvm()\n" +
-                        "}");
-
-        Files.writeString(quarkusBuild,
-                "plugins {\n" +
-                        "    id(\"io.quarkus\")\n" +
-                        "}\n" +
-                        "\n" +
-                        "repositories {\n" +
-                        "    mavenCentral()\n" +
-                        "}\n" +
-                        "\n" +
-                        "dependencies {\n" +
-                        "    implementation(project(\":mpp\"))\n" +
-                        "}");
-
-        BuildResult result = GradleRunner.create()
-                .withPluginClasspath()
-                .withProjectDir(testProjectDir.toFile())
-                .withArguments("quarkusGenerateCode")
-                .build();
-
-        assertEquals(SUCCESS, result.task(":quarkus:quarkusGenerateCode").getOutcome());
-    }
-
-    private static List<String> getDependantProvidedTaskName(Task task) {
+    private static final List<String> getDependantProvidedTaskName(Task task) {
         List<String> dependantTaskNames = new ArrayList<>();
         for (Object t : task.getDependsOn()) {
             try {

--- a/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/QuarkusPluginTest.java
+++ b/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/QuarkusPluginTest.java
@@ -1,10 +1,15 @@
 package io.quarkus.gradle;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.gradle.testkit.runner.TaskOutcome.SUCCESS;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
@@ -16,11 +21,17 @@ import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.TaskContainer;
 import org.gradle.testfixtures.ProjectBuilder;
+import org.gradle.testkit.runner.BuildResult;
+import org.gradle.testkit.runner.GradleRunner;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 
 import io.quarkus.gradle.extension.QuarkusPluginExtension;
 
 public class QuarkusPluginTest {
+
+    @TempDir
+    Path testProjectDir;
 
     @Test
     public void shouldCreateTasks() {
@@ -67,7 +78,7 @@ public class QuarkusPluginTest {
     }
 
     @Test
-    public void shouldReturnMutlipleOutputSourceDirectories() {
+    public void shouldReturnMultipleOutputSourceDirectories() {
         Project project = ProjectBuilder.builder().build();
         project.getPluginManager().apply(QuarkusPlugin.ID);
         project.getPluginManager().apply("java");
@@ -82,6 +93,67 @@ public class QuarkusPluginTest {
                 new File(project.getBuildDir(), "classes/scala/main"),
                 new File(project.getBuildDir(), "classes/scala/test"));
 
+    }
+
+    @Test
+    public void shouldNotFailOnProjectDependenciesWithoutMain() throws IOException {
+        var kotlinVersion = System.getProperty("kotlin_version", "1.7.20");
+        var settingFile = testProjectDir.resolve("settings.gradle.kts");
+        var mppProjectDir = testProjectDir.resolve("mpp");
+        var quarkusProjectDir = testProjectDir.resolve("quarkus");
+        var mppBuild = mppProjectDir.resolve("build.gradle.kts");
+        var quarkusBuild = quarkusProjectDir.resolve("build.gradle.kts");
+        Files.createDirectory(mppProjectDir);
+        Files.createDirectory(quarkusProjectDir);
+        Files.writeString(settingFile,
+                "rootProject.name = \"quarkus-mpp-sample\"\n" +
+                        "\n" +
+                        "include(\n" +
+                        "    \"mpp\",\n" +
+                        "    \"quarkus\"\n" +
+                        ")");
+
+        Files.writeString(mppBuild,
+                "buildscript {\n" +
+                        "    repositories {\n" +
+                        "        mavenLocal()\n" +
+                        "        mavenCentral()\n" +
+                        "    }\n" +
+                        "    dependencies {\n" +
+                        "        classpath(\"org.jetbrains.kotlin:kotlin-gradle-plugin:" + kotlinVersion + "\")\n" +
+                        "    }\n" +
+                        "}\n" +
+                        "\n" +
+                        "apply(plugin = \"org.jetbrains.kotlin.multiplatform\")\n" +
+                        "\n" +
+                        "repositories {\n" +
+                        "    mavenCentral()\n" +
+                        "}\n" +
+                        "\n" +
+                        "configure<org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension>{\n" +
+                        "    jvm()\n" +
+                        "}");
+
+        Files.writeString(quarkusBuild,
+                "plugins {\n" +
+                        "    id(\"io.quarkus\")\n" +
+                        "}\n" +
+                        "\n" +
+                        "repositories {\n" +
+                        "    mavenCentral()\n" +
+                        "}\n" +
+                        "\n" +
+                        "dependencies {\n" +
+                        "    implementation(project(\":mpp\"))\n" +
+                        "}");
+
+        BuildResult result = GradleRunner.create()
+                .withPluginClasspath()
+                .withProjectDir(testProjectDir.toFile())
+                .withArguments("quarkusGenerateCode")
+                .build();
+
+        assertEquals(SUCCESS, result.task(":quarkus:quarkusGenerateCode").getOutcome());
     }
 
     private static final List<String> getDependantProvidedTaskName(Task task) {

--- a/devtools/gradle/gradle-extension-plugin/src/main/java/io/quarkus/extension/gradle/dependency/DeploymentClasspathBuilder.java
+++ b/devtools/gradle/gradle-extension-plugin/src/main/java/io/quarkus/extension/gradle/dependency/DeploymentClasspathBuilder.java
@@ -43,7 +43,9 @@ public class DeploymentClasspathBuilder {
                             dependencies);
                 } else {
                     DependencyUtils.requireDeploymentDependency(deploymentConfigurationName, extension, dependencies);
-                    alreadyProcessed.add(extension.getExtensionId());
+                    if (!alreadyProcessed.add(extension.getExtensionId())) {
+                        continue;
+                    }
                 }
             }
         });
@@ -71,9 +73,9 @@ public class DeploymentClasspathBuilder {
         }
         Set<ExtensionDependency> extensions = new LinkedHashSet<>();
         for (ResolvedArtifact moduleArtifact : dependency.getModuleArtifacts()) {
-            var optionalExtension = DependencyUtils.getOptionalExtensionInfo(project, moduleArtifact);
-            if (optionalExtension.isPresent()) {
-                extensions.add(optionalExtension.get());
+            ExtensionDependency extension = DependencyUtils.getExtensionInfoOrNull(project, moduleArtifact);
+            if (extension != null) {
+                extensions.add(extension);
                 return extensions;
             }
         }

--- a/devtools/gradle/gradle-extension-plugin/src/main/java/io/quarkus/extension/gradle/tasks/ValidateExtensionTask.java
+++ b/devtools/gradle/gradle-extension-plugin/src/main/java/io/quarkus/extension/gradle/tasks/ValidateExtensionTask.java
@@ -2,9 +2,7 @@ package io.quarkus.extension.gradle.tasks;
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Optional;
 import java.util.Set;
-import java.util.stream.Collectors;
 
 import javax.inject.Inject;
 
@@ -20,6 +18,7 @@ import org.gradle.api.tasks.TaskAction;
 import io.quarkus.bootstrap.model.AppArtifactKey;
 import io.quarkus.extension.gradle.QuarkusExtensionConfiguration;
 import io.quarkus.gradle.tooling.dependency.DependencyUtils;
+import io.quarkus.gradle.tooling.dependency.ExtensionDependency;
 
 public class ValidateExtensionTask extends DefaultTask {
 
@@ -63,7 +62,13 @@ public class ValidateExtensionTask extends DefaultTask {
                 deploymentModuleKeys);
         deploymentModuleKeys.removeAll(existingDeploymentModuleKeys);
 
-        boolean hasErrors = !invalidRuntimeArtifacts.isEmpty() || !deploymentModuleKeys.isEmpty();
+        boolean hasErrors = false;
+        if (!invalidRuntimeArtifacts.isEmpty()) {
+            hasErrors = true;
+        }
+        if (!deploymentModuleKeys.isEmpty()) {
+            hasErrors = true;
+        }
 
         if (hasErrors) {
             printValidationErrors(invalidRuntimeArtifacts, deploymentModuleKeys);
@@ -71,13 +76,15 @@ public class ValidateExtensionTask extends DefaultTask {
     }
 
     private List<AppArtifactKey> collectRuntimeExtensionsDeploymentKeys(Set<ResolvedArtifact> runtimeArtifacts) {
-        return runtimeArtifacts.stream()
-                .map(resolvedArtifact -> DependencyUtils.getOptionalExtensionInfo(getProject(), resolvedArtifact))
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .map(extension -> new AppArtifactKey(extension.getDeploymentModule().getGroupId(),
-                        extension.getDeploymentModule().getArtifactId()))
-                .collect(Collectors.toList());
+        List<AppArtifactKey> runtimeExtensions = new ArrayList<>();
+        for (ResolvedArtifact resolvedArtifact : runtimeArtifacts) {
+            ExtensionDependency extension = DependencyUtils.getExtensionInfoOrNull(getProject(), resolvedArtifact);
+            if (extension != null) {
+                runtimeExtensions.add(new AppArtifactKey(extension.getDeploymentModule().getGroupId(),
+                        extension.getDeploymentModule().getArtifactId()));
+            }
+        }
+        return runtimeExtensions;
     }
 
     private List<AppArtifactKey> findExtensionInConfiguration(Set<ResolvedArtifact> deploymentArtifacts,

--- a/devtools/gradle/gradle-model/build.gradle
+++ b/devtools/gradle/gradle-model/build.gradle
@@ -3,7 +3,6 @@ dependencies {
     implementation "io.quarkus:quarkus-bootstrap-gradle-resolver:${version}"
     implementation "org.jetbrains.kotlin:kotlin-gradle-plugin-api:${kotlin_version}"
     testImplementation "io.quarkus:quarkus-devtools-testing:${version}"
-    testImplementation "org.jetbrains.kotlin:kotlin-gradle-plugin:${kotlin_version}"
 }
 
 task sourcesJar(type: Jar, dependsOn: classes) {

--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java
@@ -6,10 +6,7 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 import java.util.Set;
-import java.util.function.Function;
-import java.util.stream.Collectors;
 
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
@@ -40,7 +37,8 @@ public class ConditionalDependenciesEnabler {
     private final Set<ArtifactKey> existingArtifacts = new HashSet<>();
     private final List<Dependency> unsatisfiedConditionalDeps = new ArrayList<>();
 
-    public ConditionalDependenciesEnabler(Project project, LaunchMode mode, Configuration platforms) {
+    public ConditionalDependenciesEnabler(Project project, LaunchMode mode,
+            Configuration platforms) {
         this.project = project;
         this.enforcedPlatforms = platforms;
 
@@ -61,7 +59,7 @@ public class ConditionalDependenciesEnabler {
                     final Dependency conditionalDep = unsatisfiedConditionalDeps.get(i);
                     // Try to resolve it with the latest evolved graph available
                     if (resolveConditionalDependency(conditionalDep)) {
-                        // Mark the resolution as a success, so we know the graph has evolved
+                        // Mark the resolution as a success so we know the graph evolved
                         satisfiedConditionalDeps = true;
                         unsatisfiedConditionalDeps.remove(i);
                     } else {
@@ -90,25 +88,23 @@ public class ConditionalDependenciesEnabler {
     }
 
     private void collectConditionalDependencies(Set<ResolvedArtifact> runtimeArtifacts) {
-        addToMasterList(runtimeArtifacts);
-        var artifactExtensions = getArtifactExtensions(runtimeArtifacts);
-        allExtensions.putAll(artifactExtensions);
-        artifactExtensions.forEach((ignored, extension) -> queueAbsentExtensionConditionalDependencies(extension));
-    }
-
-    private void addToMasterList(Set<ResolvedArtifact> artifacts) {
-        artifacts.stream().map(ConditionalDependenciesEnabler::getKey).forEach(existingArtifacts::add);
-    }
-
-    private Map<ModuleVersionIdentifier, ExtensionDependency> getArtifactExtensions(Set<ResolvedArtifact> runtimeArtifacts) {
-        return runtimeArtifacts.stream()
-                .flatMap(artifact -> DependencyUtils.getOptionalExtensionInfo(project, artifact).stream())
-                .collect(Collectors.toMap(ExtensionDependency::getExtensionId, Function.identity()));
-    }
-
-    private void queueAbsentExtensionConditionalDependencies(ExtensionDependency extension) {
-        extension.getConditionalDependencies().stream().filter(dep -> !exists(dep))
-                .forEach(dep -> queueConditionalDependency(extension, dep));
+        // For every artifact in the dependency graph:
+        for (ResolvedArtifact artifact : runtimeArtifacts) {
+            // Add to master list of artifacts:
+            existingArtifacts.add(getKey(artifact));
+            ExtensionDependency extension = DependencyUtils.getExtensionInfoOrNull(project, artifact);
+            // If this artifact represents an extension:
+            if (extension != null) {
+                // Add to master list of accepted extensions:
+                allExtensions.put(extension.getExtensionId(), extension);
+                for (Dependency conditionalDep : extension.getConditionalDependencies()) {
+                    // If the dependency is not present yet in the graph, queue it for resolution later
+                    if (!exists(conditionalDep)) {
+                        queueConditionalDependency(extension, conditionalDep);
+                    }
+                }
+            }
+        }
     }
 
     private boolean resolveConditionalDependency(Dependency conditionalDep) {
@@ -116,31 +112,51 @@ public class ConditionalDependenciesEnabler {
         final Configuration conditionalDeps = createConditionalDependenciesConfiguration(project, conditionalDep);
         Set<ResolvedArtifact> resolvedArtifacts = conditionalDeps.getResolvedConfiguration().getResolvedArtifacts();
 
-        boolean isConditionalDependencyResolved = resolvedArtifacts.stream()
-                .filter(artifact -> areEquals(conditionalDep, artifact))
-                .flatMap(artifact -> DependencyUtils.getOptionalExtensionInfo(project, artifact).stream())
-                .filter(extension -> extension.getDependencyConditions().isEmpty()
-                        || exist(extension.getDependencyConditions()))
-                .findFirst().map(extension -> {
-                    enableConditionalDependency(extension.getExtensionId());
-                    return true;
-                }).orElse(false);
-
-        if (isConditionalDependencyResolved) {
-            addToMasterList(resolvedArtifacts);
-            var artifactExtensions = getArtifactExtensions(resolvedArtifacts);
-            artifactExtensions.forEach((id, extension) -> extension.setConditional(true));
-            allExtensions.putAll(artifactExtensions);
-            artifactExtensions.forEach((ignored, extension) -> queueAbsentExtensionConditionalDependencies(extension));
+        boolean satisfied = false;
+        // Resolved artifacts don't have great linking back to the original artifact, so I think
+        // this loop is trying to find the artifact that represents the original conditional
+        // dependency
+        for (ResolvedArtifact artifact : resolvedArtifacts) {
+            if (conditionalDep.getName().equals(artifact.getName())
+                    && conditionalDep.getVersion().equals(artifact.getModuleVersion().getId().getVersion())
+                    && artifact.getModuleVersion().getId().getGroup().equals(conditionalDep.getGroup())) {
+                // Once the dependency is found, reload the extension info from within
+                final ExtensionDependency extensionDependency = DependencyUtils.getExtensionInfoOrNull(project, artifact);
+                // Now check if this conditional dependency is resolved given the latest graph evolution
+                if (extensionDependency != null && (extensionDependency.getDependencyConditions().isEmpty()
+                        || exist(extensionDependency.getDependencyConditions()))) {
+                    satisfied = true;
+                    enableConditionalDependency(extensionDependency.getExtensionId());
+                    break;
+                }
+            }
         }
 
-        return isConditionalDependencyResolved;
-    }
+        // No resolution (yet); give up.
+        if (!satisfied) {
+            return false;
+        }
 
-    private boolean areEquals(Dependency dependency, ResolvedArtifact artifact) {
-        return dependency.getName().equals(artifact.getName())
-                && Objects.equals(dependency.getVersion(), artifact.getModuleVersion().getId().getVersion())
-                && artifact.getModuleVersion().getId().getGroup().equals(dependency.getGroup());
+        // The conditional dependency resolved! Let's now add all of /its/ dependencies
+        for (ResolvedArtifact artifact : resolvedArtifacts) {
+            // First add the artifact to the master list
+            existingArtifacts.add(getKey(artifact));
+            ExtensionDependency extensionDependency = DependencyUtils.getExtensionInfoOrNull(project, artifact);
+            if (extensionDependency == null) {
+                continue;
+            }
+            // If this artifact represents an extension, mark this one as a conditional extension
+            extensionDependency.setConditional(true);
+            // Add to the master list of accepted extensions
+            allExtensions.put(extensionDependency.getExtensionId(), extensionDependency);
+            for (Dependency cd : extensionDependency.getConditionalDependencies()) {
+                // Add any unsatisfied/unresolved conditional dependencies of this dependency to the queue
+                if (!exists(cd)) {
+                    queueConditionalDependency(extensionDependency, cd);
+                }
+            }
+        }
+        return satisfied;
     }
 
     private void queueConditionalDependency(ExtensionDependency extension, Dependency conditionalDep) {

--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/dependency/DependencyUtils.java
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/dependency/DependencyUtils.java
@@ -10,7 +10,6 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
-import java.util.Optional;
 import java.util.Properties;
 
 import org.gradle.api.GradleException;
@@ -74,59 +73,48 @@ public class DependencyUtils {
         return String.join(":", artifactCoords.getGroupId(), artifactCoords.getArtifactId(), artifactCoords.getVersion());
     }
 
-    public static Optional<ExtensionDependency> getOptionalExtensionInfo(Project project, ResolvedArtifact artifact) {
-        return loadExtensionDependencyFromProject(artifact, project)
-                .or(() -> loadExtensionDependencyFromDir(artifact, project))
-                .or(() -> loadExtensionDependencyFromJar(artifact, project));
-    }
+    public static ExtensionDependency getExtensionInfoOrNull(Project project, ResolvedArtifact artifact) {
+        ModuleVersionIdentifier artifactId = artifact.getModuleVersion().getId();
+        File artifactFile = artifact.getFile();
 
-    private static Optional<ExtensionDependency> loadExtensionDependencyFromProject(ResolvedArtifact artifact,
-            Project project) {
-        Optional<Project> projectDep = Optional.of(artifact.getId().getComponentIdentifier())
-                .filter(ProjectComponentIdentifier.class::isInstance)
-                .map(ProjectComponentIdentifier.class::cast)
-                .map(ProjectComponentIdentifier::getProjectPath)
-                .map(projectPath -> project.getRootProject().findProject(projectPath));
+        if (artifact.getId().getComponentIdentifier() instanceof ProjectComponentIdentifier) {
+            final Project projectDep = project.getRootProject().findProject(
+                    ((ProjectComponentIdentifier) artifact.getId().getComponentIdentifier()).getProjectPath());
+            SourceSetContainer sourceSets = projectDep == null ? null
+                    : projectDep.getExtensions().findByType(SourceSetContainer.class);
+            if (sourceSets != null) {
+                SourceSet mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                File resourcesDir = mainSourceSet.getOutput().getResourcesDir();
+                Path descriptorPath = resourcesDir.toPath().resolve(BootstrapConstants.DESCRIPTOR_PATH);
+                if (Files.exists(descriptorPath)) {
+                    return loadExtensionInfo(project, descriptorPath, artifactId, projectDep);
+                }
+            }
+        }
 
-        return projectDep
-                .map(Project::getExtensions)
-                .map(container -> container.findByType(SourceSetContainer.class))
-                .map(container -> container.findByName(SourceSet.MAIN_SOURCE_SET_NAME))
-                .map(it -> it.getOutput().getResourcesDir())
-                .map(File::toPath)
-                .flatMap(resourceDir -> loadOptionalExtensionInfo(project, resourceDir, artifact.getModuleVersion().getId(),
-                        projectDep.get()));
-    }
-
-    private static Optional<ExtensionDependency> loadExtensionDependencyFromDir(ResolvedArtifact artifact, Project project) {
-        return Optional.of(artifact.getFile().toPath()).filter(Files::exists)
-                .flatMap(path -> loadOptionalExtensionInfo(project, path, artifact.getModuleVersion().getId(), null));
-    }
-
-    private static Optional<ExtensionDependency> loadExtensionDependencyFromJar(ResolvedArtifact artifact, Project project) {
-        return Optional.of(artifact)
-                .filter(it -> ArtifactCoords.TYPE_JAR.equals(it.getExtension()))
-                .filter(it -> Files.exists(it.getFile().toPath()))
-                .flatMap(it -> {
-                    try (FileSystem artifactFs = ZipUtils.newFileSystem(it.getFile().toPath())) {
-                        return loadOptionalExtensionInfo(project, artifactFs.getPath(""), artifact.getModuleVersion().getId(),
-                                null);
-                    } catch (IOException e) {
-                        throw new GradleException("Failed to read " + it.getFile(), e);
-                    }
-                });
-    }
-
-    private static Optional<ExtensionDependency> loadOptionalExtensionInfo(Project project, Path resourcePath,
-            ModuleVersionIdentifier extensionId, Project extensionProject) {
-        return Optional.of(resourcePath)
-                .map(path -> path.resolve(BootstrapConstants.DESCRIPTOR_PATH))
-                .filter(Files::exists)
-                .map(descriptorPath -> loadExtensionInfo(project, descriptorPath, extensionId, extensionProject));
+        if (!artifactFile.exists()) {
+            return null;
+        }
+        if (artifactFile.isDirectory()) {
+            Path descriptorPath = artifactFile.toPath().resolve(BootstrapConstants.DESCRIPTOR_PATH);
+            if (Files.exists(descriptorPath)) {
+                return loadExtensionInfo(project, descriptorPath, artifactId, null);
+            }
+        } else if (ArtifactCoords.TYPE_JAR.equals(artifact.getExtension())) {
+            try (FileSystem artifactFs = ZipUtils.newFileSystem(artifactFile.toPath())) {
+                Path descriptorPath = artifactFs.getPath(BootstrapConstants.DESCRIPTOR_PATH);
+                if (Files.exists(descriptorPath)) {
+                    return loadExtensionInfo(project, descriptorPath, artifactId, null);
+                }
+            } catch (IOException e) {
+                throw new GradleException("Failed to read " + artifactFile, e);
+            }
+        }
+        return null;
     }
 
     private static ExtensionDependency loadExtensionInfo(Project project, Path descriptorPath,
-            ModuleVersionIdentifier extensionId, Project extensionProject) {
+            ModuleVersionIdentifier exentionId, Project extensionProject) {
         final Properties extensionProperties = new Properties();
         try (BufferedReader reader = Files.newBufferedReader(descriptorPath)) {
             extensionProperties.load(reader);
@@ -150,10 +138,10 @@ public class DependencyUtils {
         final ArtifactKey[] constraints = BootstrapUtils
                 .parseDependencyCondition(extensionProperties.getProperty(BootstrapConstants.DEPENDENCY_CONDITION));
         if (extensionProject != null) {
-            return new LocalExtensionDependency(extensionProject, extensionId, deploymentModule, conditionalDependencies,
+            return new LocalExtensionDependency(extensionProject, exentionId, deploymentModule, conditionalDependencies,
                     constraints == null ? Collections.emptyList() : Arrays.asList(constraints));
         }
-        return new ExtensionDependency(extensionId, deploymentModule, conditionalDependencies,
+        return new ExtensionDependency(exentionId, deploymentModule, conditionalDependencies,
                 constraints == null ? Collections.emptyList() : Arrays.asList(constraints));
     }
 

--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/dependency/DependencyUtils.java
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/dependency/DependencyUtils.java
@@ -83,7 +83,10 @@ public class DependencyUtils {
             SourceSetContainer sourceSets = projectDep == null ? null
                     : projectDep.getExtensions().findByType(SourceSetContainer.class);
             if (sourceSets != null) {
-                SourceSet mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                SourceSet mainSourceSet = sourceSets.findByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                if (mainSourceSet == null) {
+                    return null;
+                }
                 File resourcesDir = mainSourceSet.getOutput().getResourcesDir();
                 Path descriptorPath = resourcesDir.toPath().resolve(BootstrapConstants.DESCRIPTOR_PATH);
                 if (Files.exists(descriptorPath)) {

--- a/integration-tests/gradle/pom.xml
+++ b/integration-tests/gradle/pom.xml
@@ -88,6 +88,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-validator</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jdbc-h2</artifactId>
         </dependency>
         <dependency>
@@ -113,6 +117,14 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-resteasy-jackson</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-reactive</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-reactive-jackson</artifactId>
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
@@ -219,6 +231,19 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-validator-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jdbc-h2-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
@@ -272,6 +297,32 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-resteasy-jackson-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-reactive-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-reactive-jackson-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
             <scope>test</scope>

--- a/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/build.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/build.gradle.kts
@@ -1,0 +1,49 @@
+plugins {
+    kotlin("jvm") version "1.7.20"
+    kotlin("plugin.allopen") version "1.7.20"
+    id("io.quarkus")
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex("io.quarkus.*")
+        }
+    }
+    mavenCentral()
+}
+
+val quarkusPlatformGroupId: String by project
+val quarkusPlatformArtifactId: String by project
+val quarkusPlatformVersion: String by project
+
+dependencies {
+    implementation(enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}"))
+    implementation("io.quarkus:quarkus-resteasy-reactive-jackson")
+    implementation("io.quarkus:quarkus-resteasy-reactive")
+    implementation("io.quarkus:quarkus-kotlin")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+    implementation("io.quarkus:quarkus-arc")
+    implementation("io.quarkus:quarkus-hibernate-validator")
+    testImplementation("io.quarkus:quarkus-junit5")
+    testImplementation("io.rest-assured:rest-assured")
+}
+
+group = "org.acme"
+version = "1.0.0-SNAPSHOT"
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_11
+    targetCompatibility = JavaVersion.VERSION_11
+}
+
+allOpen {
+    annotation("javax.ws.rs.Path")
+    annotation("javax.enterprise.context.ApplicationScoped")
+    annotation("io.quarkus.test.junit.QuarkusTest")
+}
+
+tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
+    kotlinOptions.jvmTarget = JavaVersion.VERSION_11.toString()
+    kotlinOptions.javaParameters = true
+}

--- a/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/gradle.properties
@@ -1,0 +1,2 @@
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus

--- a/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/settings.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/settings.gradle.kts
@@ -1,0 +1,16 @@
+pluginManagement {
+    val quarkusPluginVersion: String by settings
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex("io.quarkus.*")
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+        id("io.quarkus") version quarkusPluginVersion
+    }
+}
+rootProject.name="code-with-quarkus"

--- a/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/src/main/kotlin/org/acme/AdminResource.kt
+++ b/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/src/main/kotlin/org/acme/AdminResource.kt
@@ -1,0 +1,16 @@
+package org.acme
+
+import org.acme.api.AdminAPI
+import javax.enterprise.context.ApplicationScoped
+import javax.ws.rs.core.Response
+import javax.ws.rs.core.UriInfo
+
+
+@ApplicationScoped
+class AdminResource : AdminAPI {
+
+  override suspend fun createAccount(tenantId: String, body: String, uriInfo: UriInfo): Response {
+    return Response.noContent().build()
+  }
+
+}

--- a/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/src/main/kotlin/org/acme/api/API.kt
+++ b/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/src/main/kotlin/org/acme/api/API.kt
@@ -1,0 +1,22 @@
+package org.acme.api
+
+import javax.validation.Valid
+import javax.ws.rs.*
+import javax.ws.rs.core.Context
+import javax.ws.rs.core.Response
+import javax.ws.rs.core.UriInfo
+
+@Path(value = "/admin/api/v1.0")
+@Produces(value = ["application/json"])
+@Consumes(value = ["application/json"])
+interface AdminAPI {
+
+  @POST
+  @Path(value = "/tenants/{tenantId}/accounts")
+  suspend fun createAccount(
+    @PathParam(value = "tenantId") tenantId: String,
+    body: String,
+    @Context uriInfo: UriInfo
+  ): Response
+
+}

--- a/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/src/test/kotlin/org/acme/AdminResourceTest.kt
+++ b/integration-tests/gradle/src/main/resources/conditional-dependencies-kotlin/src/test/kotlin/org/acme/AdminResourceTest.kt
@@ -1,0 +1,23 @@
+package org.acme
+
+import io.quarkus.test.junit.QuarkusTest
+import io.restassured.RestAssured.given
+import org.hamcrest.CoreMatchers.`is`
+import org.junit.jupiter.api.Test
+
+@QuarkusTest
+class AdminResourceTest {
+
+    @Test
+    fun testHelloEndpoint() {
+        given()
+          .`when`()
+          .contentType("application/json")
+          .body("{}")
+          .post("/admin/api/v1.0/tenants/1/accounts")
+          .then()
+          .log().ifValidationFails()
+          .statusCode(204)
+    }
+
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/ConditionalDependenciesKotlinTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/ConditionalDependenciesKotlinTest.java
@@ -1,0 +1,17 @@
+package io.quarkus.gradle;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+import org.junit.jupiter.api.Test;
+
+public class ConditionalDependenciesKotlinTest extends QuarkusGradleWrapperTestBase {
+
+    @Test
+    public void buildProject() throws IOException, URISyntaxException, InterruptedException {
+        final File projectDir = getProjectDir("conditional-dependencies-kotlin");
+
+        runGradleWrapper(projectDir, "clean", "build");
+    }
+}

